library/alloc/src/string.rs

```diff
@@ -2224,6 +2224,62 @@ impl ToString for char {
     }
 }
 
+#[stable(feature = "u8_to_string_specialization", since = "1.999.0")]
+impl ToString for u8 {
+    #[inline]
+    fn to_string(&self) -> String {
+        let mut result = String::with_capacity(3);
+        let mut n = *self;
+        if n >= 100 {
+            result.push((b'0' + n / 100) as char);
+            n %= 100;
+        }
+        if !result.is_empty() || n >= 10 {
+            result.push((b'0' + n / 10) as char);
+            n %= 10;
+        };
+        result.push((b'0' + n) as char);
+        result
+    }
+}
+
+// 2 digit decimal look up table
+static DEC_DIGITS_LUT: &[u8; 200] = b"0001020304050607080910111213141516171819\
+      2021222324252627282930313233343536373839\
+      4041424344454647484950515253545556575859\
+      6061626364656667686970717273747576777879\
+      8081828384858687888990919293949596979899";
+
+#[stable(feature = "i8_to_string_specialization", since = "1.999.0")]
+impl ToString for i8 {
+    #[inline]
+    fn to_string(&self) -> String {
+        let mut n = *self;
+        let mut vec: Vec<u8> = if n < 0 {
+            // convert the negative num to positive by summing 1 to it's 2 complement
+            // ( -128u8.abs() would panic )
+            n = (!n).wrapping_add(1);
+            let mut v = Vec::with_capacity(4);
+            v.push(b'-');
+            v
+        } else {
+            Vec::with_capacity(3)
+        };
+        let mut n = n as u8;
+        if n >= 10 {
+            if n >= 100 {
+                n -= 100;
+                vec.push(b'1');
+            }
+            let nn = n * 2;
+            vec.extend_from_slice(&DEC_DIGITS_LUT[nn as usize..=nn as usize + 1]);
+        } else {
+            vec.push(b'0' + (n as u8));
+        }
+        unsafe { String::from_utf8_unchecked(vec) }
```

Comment:
A safety comment explaining why this can't cause UB would be nice.

Reply:
Have added one in. I threw in the two debug_asserts that prove that it is safe. Idk if that was overkill?

Reply:
> // SAFETY: Vec only contains ascii so valid utf8

Why does the Vec only contain ASCII bytes though? "All bytes written to the `vec` are valid ASCII bytes and all positions of it, which have not been written to, have been removed by the `truncate` call. Therefore, `vec` only contains ASCII, which is automatically valid UTF-8."
(Yes, it does get initialized with zeros, which are also valid UTF-8 bytes, but you could have initialized with 255s and this explanation would still cover it.)

Then, instead of the `debug_assert`s, some comments explaining why these are valid ASCII bytes would complement the safety comment. Like:
```rust
// Since n < 100 at this point, and b'0' == 48, the sum is in the range 48..=57,
// representing the characters 0-9 in ASCII.
vec[free] = b'0' + n / 10;
```

Reply:
Resolving this one as we're not using unsafe.